Best practices for local development (#203)
langium/langium: CONTRIBUTING.md

## Prerequisites

> Developing Langium only requires `npm@^7.7.0`. For this you will need `Node.js 10` or greater.

Reviewer: We should check the actual minimum NodeJS version (see #192). I noticed that we use more recent APIs in some places, so NodeJS 10 certainly doesn't work, but that's fine because it's old and no longer supported.

Looking at https://nodejs.org/en/about/releases/ I think we should use version 12 as minimum version.

Author: Ok, so minimum version is Maintenance LTS. Are there further requirements apart from workspaces that we have on a specific npm version?

Reviewer: No, the workspaces feature is what we need from npm, and regarding NodeJS we need to be careful not to use API that is only available in versions > 12.

> ## Best practices for local development

Reviewer: We should start the contributing guide with the instructions to develop the packages in our repository, and make that as simple as possible, to avoid creating the impression that we somehow need a complex setup. The explanations you added here are useful, but I'd put them into an "advanced" section to clarify that it's not always necessary.

Author: I agree. I was very verbose on a specific aspect that is maybe only relevant for 10% of future users.
The guide should be simple for the normal use-case which is straightforward and it should make this more obvious (re-structure, simple first, advance use case later).

Langium itself contains a couple of language examples. If you want to write your own grammar you should use `yo langium` and create an independent project. If you just want to use Langium out of the box, then you are ready to start and don't need to necessarily read further. But if for whatever reason you want to extend Langium locally and you want to use your changes in your own language project then you should follow some recommendations.

### npm dependency resolution pitfalls

> When you add a dependency to a `package.json` npm resolves this from the configured registry (usually https://registry.npmjs.org/). In other dependency management system you can create something like a dev or snapshot version and install it locally, but this is not possible as such with npm. If you want to achieve a similar behavior you need to link your packages globally (global refers to you user's account scope, local refers to the project scope). When you do this you forcefully overwrite what was downloaded from the npm registry. Incrementing the version of Langium itself will led to problems, because npm will not be able to resolve those from its registry and therefore fail any install attempt in projects referring to Langium.

Reviewer: ```suggestion
When you add a dependency to a `package.json` npm resolves this from the configured registry (usually https://registry.npmjs.org/). In other dependency management systems you can create something like a dev or snapshot version and install it locally, but this is not possible as such with npm. If you want to achieve a similar behavior you need to link your packages globally (global refers to you user's account scope, local refers to the project scope). When you do this you forcefully overwrite what was downloaded from the npm registry. Incrementing the version of Langium itself will lead to problems, because npm will not be able to resolve those from its registry and therefore fail any install attempt in projects referring to Langium.
```

Author: I have already adjusted the test locally and it will be part of an upcoming commit.

> Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`). If you execute `npm install` on the top-level of the project all dependencies will be gathered locally inside `node_modules` and all four packages or sub-projects will be build. **Warning**: Do not run npm install directly inside the four Langium packages as it will mess up dependency resolution once you link packages globally.

Reviewer: ```suggestion
Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`). If you execute `npm install` on the top-level of the project all dependencies will be gathered locally inside `node_modules` and all four packages or sub-projects will be built. **Warning**: Do not run npm install directly inside the four Langium packages as it will mess up dependency resolution once you link packages globally.
```

Author: I have already adjusted the test locally and it will be part of an upcoming commit.

## Prerequisites

For developing Langium you require at least the Maintenance LTS version of Node.js (currently 12) and at least npm version 7.7.0 (`npm@^7.7.0`) to be able to use npm workspaces. With Node.js versions below 16 you have to install the correct version of npm with `npm install -g npm@7`.

## Local Development Guide

We outline three use-cases:
1. Developing Langium itself

> 2. Developing you own language and use Langium as tool

Author: I will change with next commit.

> 3. Developing Langium itself and developing your own language itself.

Author: Yes, sorry it should be "in parallel". I will change with next commit.

The first two use-cases are the common ones. The third one is less common, but there are some recommendations you should know.

### Developing Langium

> Development of Langium does not necessarily require VSCode (https://code.visualstudio.com/download), but without it you will not be able to use language server and its extension for VSCode.

Reviewer: ```suggestion
Development of Langium does not necessarily require VSCode (https://code.visualstudio.com/download), but without it you will not be able to use the language server and its extension for VSCode.
```

Author: I will change it with next commit.

> Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`).

Reviewer: Writing details such as the number and the names of contained packages here means we have to maintain this whenever there are changes. I'd rather shorten this to "multiple packages", and maybe link that to the `packages` folder in GitHub.

Author: Good idea. I will change it accordingly.

Once you cloned this repository, perform `npm install` and afterwards `npm run watch` either from your terminal of choice or from a terminal within VSCode.  The second command runs the TypeScript compiler (`tsc`) in watch mode which generates new JavaScript code every time you save a TypeScript file in one of the Langium packages.

Take a look at the examples readily available inside the repository to get started:
* [Arithmetics](./examples/arithmetics/README.md)
* [Domain Model](./examples/domainmodel/README.md)
* [State Machine](./examples/statemachine/README.md)

### Developing your own language

> The Please follow the instructions outlined in the [Langium package documentation](./packages/langium/README.md) to get started.

Author: I will change it accordingly

### Developing Langium and your own language in parallel

> If for whatever reason you want:

Author: I will change it accordingly

- Develop or change Langium locally
- Use your Langium changes in your own language project (independent of Langium repository)
- Debug all code in all projects
then you should follow some recommendations.

#### npm dependency resolution pitfalls

> If you are have long experience in using node and npm this sub-chapter will likely offer no new info an you can skip it. But, if you are more familiar with other languages, build and dependency management eco-systems then this sub-chapter could be helpful to you.

Author: I will change it accordingly

```
inside `package.json`.

#### Use VSCode workspace to simplify debugging

When debugging a project that depends on another project under development using a `code-workspace` makes life easier. Open an empty VSCode instance and *"Add Folder to Workspace..."* and add the Langium directory and your language project directory to the workspace. Save it with *"Save Workspace As..."*. It is recommended to have your language project and Langium itself reside on the same level on the same directory level.

**Reminder:** Don't forget to run `npm run watch` in your language project and the Langium package or any other package where you want to change code and see the effects when executing and debugging.

In `.vscode/launch.json` of your language project add the following launch configuration:
```json
        {
			"name": "Attach",
			"port": 6009,
			"request": "attach",
			"skipFiles": [
				"<node_internals>/**"
			],
			"sourceMaps": true,
			"outFiles": [
				"${workspaceFolder}/out/**/*.js"
			],
			"type": "pwa-node"
		}
```
With the above configuration you can attach the debugger to the language server which effectively means you are able to debug your local version of Langium.

> **Hint:** If you want to halt the language server start until you attach the debugger you have to `--inspect=6009` to `--inspect-brk=6009` in the language project's `src/extension.ts` like shown below:

Reviewer: ```suggestion
**Hint:** If you want to halt the language server start until you attach the debugger you have to change `--inspect=6009` to `--inspect-brk=6009` in the language project's `src/extension.ts` like shown below:
```

Reviewer: We could control this via an environment variable so a code change is not necessary. Could you create an issue for that?

Author: I will change it accordingly and I just created #215

A project you created with `yo langium` contains a dependency to `langium` (e.g. `0.1.0`) and a dev-dependency to `langium-cli`) inside `package.json` by default. Now, the easiest way to switch to your own build is to locally reference the required Langium packages folders.
Let's assume your local Langium checkout resides on the same level as your own language project, then these two commands will set the references inside the `package.json`:
```shell
npm install ../langium/packages/langium

> npm install -D ../langium/packages/langium-cli

Reviewer 2: I figured out that can use `npm link -S langium` and `npm link -D langium-cli` to add those dependencies in the versions "installed"/"published" to the user global package cache. If a package is installed as link to a local folder via `npm link` the link target is then referenced in the `package.json`. Otherwise, the path of the package within the global package cache is used. If the particular package is not available in the global cache it will be downloaded from the public registry first.
In contrast to the statement in the `help` contents (seem to be equal for several commands) saying `--safe` is true by default, this is not the case for me. I had to state `--safe` or `-S` get the dependency added in the `package.json`.

I suggest to replace the above statements by the `npm link`-based ones as they appear to be more consistent with the preparations suggested above to me.

Reviewer 3: What does -S and -D mean?

Reviewer 2: -S: "add the package to the 'dependencies' list in the package.json" (since npm5 by default true for `npm install`)
-D: "add the package to the 'devDependencies' list in the package.json"

Author: I am pretty certain that you need the local reference for the VSCode debugger to work correctly in the VSCode workspace, but I may be in error here. We should check this before altering the description.

Author: ```shell
npm link -S langium
npm link -D langium-cli
```
creates the local file references as desired (my local build was globally installed before (machine-wide). So, this can replace the other instruction then. @sailingKieler It seems you can only omit the options in case of `npm install` and not in case of `npm link`. I can confirm your observation.
